src/v/kafka/server/audit_log_manager.h

```diff
@@ -0,0 +1,165 @@
+    /// publish messages. Consumers of this API should react accordingly, i.e.
+    /// return an error to the client.
+    template<InheritsFromOCSFBase T, typename... Args>
+    bool enqueue_audit_event(audit_event_type type, Args&&... args) {
+        if (_as.abort_requested()) {
+            /// Prevent auditing new messages when shutdown starts that way the
+            /// queue may be entirely flushed before shutdown
+            return false;
+        }
+        if (!_audit_enabled() || !is_audit_event_enabled(type)) {
+            return true;
+        }
+        return do_enqueue_audit_event(
+          std::make_unique<T>(std::forward<Args>(args)...));
+    }
+
+    /// Returns the number of items pending to be written to auditing log
+    ///
+    /// Note does not include records already sent to client
+    size_t pending_events() const { return _queue.size(); };
+
+    /// Returns true if the internal kafka client is allocated
+    ///
+    /// NOTE: Only works on shard_id{0}, use in unit tests
+    bool is_client_enabled() const;
+
+private:
+    ss::future<> drain();
+    ss::future<> pause();
+    ss::future<> resume();
+
+    bool is_audit_event_enabled(audit_event_type) const;
+    bool do_enqueue_audit_event(
+      std::unique_ptr<security::audit::ocsf_base_impl> msg);
+    void set_enabled_events();
+
+private:
+    /// Multi index container is efficent in terms of time and space, underlying
+    /// internal data structures are compact requiring only one node per
+    /// element. More info here:
+    /// https://www.boost.org/doc/libs/1_72_0/libs/multi_index/doc/performance.html
+    struct underlying_list {};
+    struct underlying_unordered_map {};
+    using underlying_t = boost::multi_index::multi_index_container<
+      std::unique_ptr<security::audit::ocsf_base_impl>,
+      boost::multi_index::indexed_by<
+        /// Sequenced list of entries
+        boost::multi_index::sequenced<boost::multi_index::tag<underlying_list>>,
+        /// Set of audit_messages using hashed representations as comparator
+        boost::multi_index::hashed_unique<
+          boost::multi_index::tag<underlying_unordered_map>,
+          boost::multi_index::const_mem_fun<
+            security::audit::ocsf_base_impl,
+            size_t,
+            &security::audit::ocsf_base_impl::key>>>>;
+
+    /// configuration options
+    config::binding<bool> _audit_enabled;
+    config::binding<std::chrono::milliseconds> _queue_drain_interval_ms;
+    config::binding<size_t> _max_queue_elements_per_shard;
+    config::binding<std::vector<ss::sstring>> _audit_event_types;
+    /// Size of bits of representation of enum as bitset is sizeof(uint8_t)
+    /// multiplied by number of bits in a byte
+    static constexpr auto enabled_set_bitlength
+      = 8 * sizeof(std::underlying_type_t<audit_event_type>);
+    std::bitset<enabled_set_bitlength> _enabled_event_types{0};
```

Comment:
the calculation of the size of the bitset doesn't look correct because it doesn't appear to be a function of the number of flags, it just seems to be the constant value 8*8=64.

one suggestion:

```c++
enum class type {
  ...,
  num_flags // must be last element
};
std::bitset<type::num_flags> flags;
```

Reply:
Yeah the intention was to make it the same size as the underlying type of the enum, the enum is an `uint8_t` so the math evaluates to `8 * 1 = 8`.

I guess its better to have the size be the exact number of possible enums though +1

Reply:
probably a good idea to make this work automatically now since there isn't a comment on the enum instructing the user to update the size of the bitset.